Support permit extension point

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Taking over https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/pull/93.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/issues/72

#### Special notes for your reviewer:

@kazukousen I put you in commit co-author, let me know in case you don't want to (I believe you won't 🙂).

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Support permit extension point
```

#### What are the benchmark results of this change?
<!--
If documentation-only or trivial, just write "N/A". Otherwise, run the following:

1. `make bench-example > before.txt` on the commit prior to your changes
2. `make bench-example > after.txt` on the commit of your changes
3. Paste the output of `benchstat before.txt after.txt` as a code block.

If you haven't yet installed benchstat, it looks like this.
```
$ go install golang.org/x/perf/cmd/benchstat@latest
```

-->
```benchstat

```

sanposhiho commented:
/cc @utam0k @Gekko0114

Gekko0114 commented on scheduler/plugin/plugin_test.go:
```diff
@@ -1086,6 +1083,97 @@ func TestReserve(t *testing.T) {
+			expectedStatusMessage: "name is bad",
+		},
+		{
+			name:                  "Error",
```
```suggestion
			name:                  "Wait",
```

Gekko0114 reviewed (commented):
Overall looks good :)

BTW, how about adding an explanation somewhere that you can't pass too big int64 values as we discussed previously?
https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/pull/93/files#r1468783879

sanposhiho commented:
First, changed the interface to -

```go
Permit(state CycleState, p proto.Pod, nodeName string) (status *Status, timeoutMilliSeconds uint32)
```

And, updated RATIONAL.md for the reason.

Also, actually we don't have to use hostfunction to return timeout, and so eliminated things around it.

Gekko0114 commented on scheduler/plugin/guest.go:
```diff
@@ -238,6 +242,21 @@ func (g *guest) unreserve(ctx context.Context) {
+	}
+
+	statusCode := int32(callStack[0] >> 32)
+	timeoutMilliSeconds := int32(callStack[0])
```
How about using uint32 here, just to prevent a too big positive number from changing to a negative number?

Gekko0114 commented:
Thanks! Looks good and left a very minor comment.

sanposhiho commented:
Updated.

Gekko0114 reviewed (approved):
Thanks! Left lgtm label for other reviewers

utam0k commented:
/lgtm
